docs/howtos/exposing-rust-components-to-kotlin.md

```diff
@@ -0,0 +1,203 @@
+
+## How to expose your Rust Component to Kotlin
+
+Welcome!
+
+It's great
+that you've built a Rust Component
+and want to expose it to Koltin,
+but we haven't written a complete guide yet.
+
+Please share your :+1:
+on [the relevant github issue](https://github.com/mozilla/application-services/issues/599)
+to let us know
+that you wanted it.
+
+In the meantime,
+here are some preliminary notes:
+
+### High-level overview.
+
+The [fxa-client](/components/fxa-client) component
+provides a useful example.
+We assume you have written
+a nice core of Rust code
+in a [./src/](/components/fxa-client/src) directory
+for your component.
+
+First, you will need to
+flatten the Rust API
+into a set of C FFI bindings,
+conventially in a [./ffi/](/components/fxa-client/ffi) directory.
+Use the [ffi_support](https://docs.rs/ffi-support/0.1.3/ffi_support/) crate
+to help make this easier,
+and consule that crate's documentation
+for tips and gotchas.
+
+Next, you will need to
+write Kotlin code
+that consumes that C FFI,
+conventionally in a [./android/](/components/fxa-client/android) directory.
+This code should use [JNA](https://github.com/java-native-access/jna)
+to load the C FFI via shared library
+and expose it
+as a nice safe ergonomic Kotlin API.
+
+It seems likely
+that we could provide
+a useful template
+and/or suite of shared utilities
```

Comment:
Sadly it's not as simple as just doing this. It's a tricky design problem for a lot of it due to supporting the non-megazorded configuration (where 'which "free this string's memory" function' you call *really* matters, and a lot of the stuff that looks boilerplatey needs to be able to free strings allocated from rust).

It probably needs some tricky generic type magic or something, but when I've played around with it I couldn't get it to work.

A template is probably doable though!

(I'm more commenting on the current state of affairs than about what the doc says, it can stay as is).